File: app/src/main/java/com/makentoshe/Solver.java
package com.makentoshe;

public class Calculator {

    private static String expression;
    private static int pos = -1;
    private static double curr_number;
    private static char curr_token;

    private static boolean isCharNumeric(char ch) {
        return (ch >= '0' && ch <= '9' || ch == '.' || ch == ',');
    }

    private static void get_next_token() {
        pos++;
        if(pos >= expression.length()) {
            curr_token = 'q';
        } else {
            switch(expression.charAt(pos)) {
                case '+': case '-': case '*': case '/': case '^': case '(': case ')': {
                    curr_token = expression.charAt(pos);
                    break;
                }
                case '0': case '1': case '2': case '3': case '4': case '5':
                case '6': case '7': case '8': case '9': case '.': {
                    curr_token = 'n';
                    int pos0 = pos;
                    while(pos < expression.length() && isCharNumeric(expression.charAt(pos)))
                        pos++;
                    try {
                        curr_number = Double.parseDouble(expression.substring(pos0, pos));
                    } catch (NumberFormatException e) {
                        curr_token = 'i';
                    }
                    pos--;
                    break;
                }
                default: {
                    //out.println("Incorrect symbol at position " + (pos+1));
                    curr_token = 'i';
                }
            }
        }
    }

    private static double prim() {
        switch (curr_token) {
            case 'n': {
                get_next_token();
                return curr_number;
            }
            case '-': {
                get_next_token();
                return -prim();
            }
            case '(': {
                get_next_token();
                double e = expr();
                get_next_token();
                return e;
            }
            default: {
                return expr();
            }
        }
    }

    private static double expr() {
        double left = term();
        while(true) {
            switch (curr_token) {
                case '+': {
                    get_next_token();
                    left += term();
                    break;
                }
                case '-': {
                    get_next_token();
                    left -= term();
                    break;
                }
                default: {
                    return left;
                }
            }
        }
    }

    private static double term() {
        double left = prim();
        while(true) {
            switch (curr_token) {
                case '*': {
                    get_next_token();
                    left *= prim();
                    break;
                }
                case '/': {
                    get_next_token();
                    left /= prim();
                    break;
                }
                case '^': {
                    get_next_token();
                    left = Math.pow(left, prim());
                    break;
                }
                default: {
                    return left;
                }
            }
        }
    }

    private static boolean check_expression() {
        expression = expression.replace(',', '.');
        expression = expression.replaceAll(" ", "");

        pos = -1;
        boolean res = true;

        if(expression.length() == 0)
            res = false;

        char prev_token = '(';
        get_next_token();
        while(curr_token != 'q') {
            if(curr_token == 'i')
                res = false;
            switch(prev_token) {
                case 'n': {
                    if(curr_token == '(') {
                        expression = expression.substring(0, pos) + '*' + expression.substring(pos);
                        pos++;
                    }
                    break;
                }
                case '+': case '-': case '*': case '/': case '^': case '(': {
                    if(curr_token == '+' || curr_token == '*' || curr_token == '/' || curr_token == '^')
                        res = false;
                    break;
                }
            }
            prev_token = curr_token;
            get_next_token();
        }
        if(prev_token == '+' || prev_token == '*' || prev_token == '/' || prev_token == '^' || prev_token == '-' || prev_token == '(')
            res = false;
        int braces = 0;
        for(int i = 0; i < expression.length(); ++i) {
            if(expression.charAt(i) == '(')
                braces++;
            else if(expression.charAt(i) == ')')
                braces--;
        }
        if(braces != 0)
            res = false;
        return res;
    }

    public static double calculate(String expr) {
        expression = expr;
        if(check_expression()) {
            pos = -1;
            get_next_token();
            return expr();
        }
        return 0;
    }

    private Calculator() {}
}

File: app/src/main/java/com/makentoshe/Solver.kt
package com.makentoshe

class Solver {
    private var expression: String = ""
    private var pos = -1
    private var currNumber = 0.0
    private var currToken = 0.toChar()

    private fun getNextToken() {
        pos++
        if (pos >= expression.length) {
            currToken = 'q'
        } else {
            when (expression[pos]) {
                '+', '-', '*', '/', '^', '(', ')' -> {
                    currToken = expression[pos]
                }
                '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.' -> {
                    currToken = 'n'
                    val pos0 = pos
                    while (pos < expression.length && expression[pos].isDigit()) pos++
                    try {
                        currNumber = expression.substring(pos0, pos).toDouble()
                    } catch (e: NumberFormatException) {
                        currToken = 'i'
                    }
                    pos--
                }
                else -> {
                    //out.println("Incorrect symbol at position " + (pos+1));
                    currToken = 'i'
                }
            }
        }
    }

    private fun prim(): Double {
        return when (currToken) {
            'n' -> {
                getNextToken()
                currNumber
            }
            '-' -> {
                getNextToken()
                -prim()
            }
            '(' -> {
                getNextToken()
                val e = expr()
                getNextToken()
                e
            }
            else -> {
                expr()
            }
        }
    }

    private fun expr(): Double {
        var left = term()
        while (true) {
            when (currToken) {
                '+' -> {
                    getNextToken()
                    left += term()
                }
                '-' -> {
                    getNextToken()
                    left -= term()
                }
                else -> {
                    return left
                }
            }
        }
    }

    private fun term(): Double {
        var left = prim()
        while (true) {
            when (currToken) {
                '*' -> {
                    getNextToken()
                    left *= prim()
                }
                '/' -> {
                    getNextToken()
                    left /= prim()
                }
                '^' -> {
                    getNextToken()
                    left = Math.pow(left, prim())
                }
                else -> {
                    return left
                }
            }
        }
    }

    private fun checkExpression(): Boolean {
        expression = expression.replace(',', '.')
        expression = expression.replace(" ", "")
        pos = -1
        var res = true
        if (expression.length == 0) res = false
        var prev_token = '('
        getNextToken()
        while (currToken != 'q') {
            if (currToken == 'i') res = false
            when (prev_token) {
                'n' -> {
                    if (currToken == '(') {
                        expression = expression.substring(
                            0,
                            pos
                        ) + '*' + expression.substring(pos)
                        pos++
                    }
                }
                '+', '-', '*', '/', '^', '(' -> {
                    if (currToken == '+' || currToken == '*' || currToken == '/' || currToken == '^') res =
                        false
                }
            }
            prev_token = currToken
            getNextToken()
        }
        if (prev_token == '+' || prev_token == '*' || prev_token == '/' || prev_token == '^' || prev_token == '-' || prev_token == '(') res =
            false
        var braces = 0
        for (i in 0 until expression.length) {
            if (expression[i] == '(') braces++ else if (expression[i] == ')') braces--
        }
        if (braces != 0) res = false
        pos = -1
        return res
    }

    fun solve(expr: String): Double {
        expression = expr
        if (checkExpression()) {
            getNextToken()
            return expr()
        }
        return 0.0
    }
}
